### PR TITLE
Avoid needless TypeRef allocation during erasure.

### PR DESCRIPTION
**Pull request by retronym**
Avoid needless TypeRef allocation during erasure.
- ThisType(some.package) need not be erased, we can let
  that prefix through unchanged and avoid churning through
  allocations.
- Sharpen the condition in `rebindInnerClass`. The answer
  to the deleted comment in the code, is "anonymous and
  local classes", which are specifically excluded in the
  new formulation.
- Finally, detect if erasure of the TypeRef is an identity and reuse
  the original.

Waste not, want not.

To expand on the first point, here is what used to happen during
erasure:

```
scala> val scalaPack = typeOf[Predef.type].prefix
scalaPack: $r.intp.global.Type = scala.type

scala> typeDeconstruct.show(scalaPack)
res19: String = ThisType(package scala)

scala> typeDeconstruct.show(erasure.scalaErasure(scalaPack))
res20: String = TypeRef(TypeSymbol(final class scala extends ))
```

Showing one step of the erasure type map:

```
scala> typeDeconstruct.show(scalaPack.asInstanceOf[SubType].underlying.typeOfThis)
res21: String = SingleType(pre = ThisType(package <root>), package scala)
```

### CODE REVIEW

**Comment by retronym** [comment]
Review by @adriaanm @odersky

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 1]
```diff
@@ -1043,7 +1043,7 @@ abstract class Erasure extends AddInterfaces
         case Literal(ct) if ct.tag == ClazzTag
                          && ct.typeValue.typeSymbol != definitions.UnitClass =>
           val erased = ct.typeValue match {
-            case TypeRef(pre, clazz, args) if clazz.isDerivedValueClass => scalaErasure.eraseNormalClassRef(pre, clazz)
+            case tr @ TypeRef(pre, clazz, args) if clazz.isDerivedValueClass => scalaErasure.eraseNormalClassRef(tr)
```
Comment (JamesIry) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: "args" is unused here, could be "_"

File: src/reflect/scala/reflect/internal/transform/Erasure.scala [hunk 2]
```diff
@@ -73,9 +73,8 @@ trait Erasure {
   // included (use pre.baseType(cls.owner)).
   //
   // This requires that cls.isClass.
-  protected def rebindInnerClass(pre: Type, cls: Symbol): Type = {
-    if (cls.owner.isClass) cls.owner.tpe_* else pre // why not cls.isNestedClass?
-  }
+  protected def rebindInnerClass(pre: Type, cls: Symbol): Type =
+    if (!cls.isTopLevel && !cls.isLocal) cls.owner.tpe_* else pre
```
Comment (JamesIry) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala]: Matter of taste, but I slightly prefer minimizing negation with "if (!(cls.isTopLevel || cls.isLocal))"

File: src/compiler/scala/tools/nsc/transform/Erasure.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: done.

File: src/reflect/scala/reflect/internal/transform/Erasure.scala (hunk 2 above)
Comment (retronym) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala]: i went one better and flipped the then/else so as not to negate.

**Comment by JamesIry** [comment]
LGTM

**Pull request merged**